python/binding/tritonserver_pybind.cc

```diff
@@ -0,0 +1,1991 @@
+      params.emplace_back(p->Ptr());
+    }
+    ThrowIfError(TRITONSERVER_ServerLoadModelWithParameters(
+        triton_object_, model_name.c_str(), params.data(), params.size()));
+  }
+
+  void UnloadModel(const std::string& model_name)
+  {
+    ThrowIfError(
+        TRITONSERVER_ServerUnloadModel(triton_object_, model_name.c_str()));
+  }
+
+  void UnloadModelAndDependents(const std::string& model_name)
+  {
+    ThrowIfError(TRITONSERVER_ServerUnloadModelAndDependents(
+        triton_object_, model_name.c_str()));
+  }
+
+  std::shared_ptr<PyMetrics> Metrics() const
+  {
+    struct TRITONSERVER_Metrics* metrics;
+    ThrowIfError(TRITONSERVER_ServerMetrics(triton_object_, &metrics));
+    return std::make_shared<PyMetrics>(metrics, true /* owned */);
+  }
+
+  void InferAsync(PyInferenceRequest& request, PyTrace& trace)
+  {
+    ThrowIfError(TRITONSERVER_ServerInferAsync(
+        triton_object_, request.Ptr(), trace.Ptr()));
+    // Ownership of the internal C object is transferred.
+    request.Release();
+    trace.Release();
+  }
+};
+
+class PyMetricFamily : public PyWrapper<struct TRITONSERVER_MetricFamily> {
+ public:
+  DESTRUCTOR_WITH_LOG(PyMetricFamily, TRITONSERVER_MetricFamilyDelete);
+
+  PyMetricFamily(
+      TRITONSERVER_MetricKind kind, const std::string& name,
+      const std::string& description)
+  {
+    TRITONSERVER_MetricFamilyNew(
+        &triton_object_, kind, name.c_str(), description.c_str());
+    owned_ = true;
+  }
+};
+
+class PyMetric : public PyWrapper<struct TRITONSERVER_Metric> {
+ public:
+  DESTRUCTOR_WITH_LOG(PyMetric, TRITONSERVER_MetricDelete);
+  PyMetric(
+      PyMetricFamily& family,
+      const std::vector<std::shared_ptr<PyParameter>>& labels)
+  {
+    std::vector<const struct TRITONSERVER_Parameter*> params;
+    for (const auto& label : labels) {
+      params.emplace_back(label->Ptr());
+    }
+    ThrowIfError(TRITONSERVER_MetricNew(
+        &triton_object_, family.Ptr(), params.data(), params.size()));
+    owned_ = true;
+  }
+
+  double Value() const
+  {
+    double val = 0;
+    ThrowIfError(TRITONSERVER_MetricValue(triton_object_, &val));
+    return val;
+  }
+
+  void Increment(double val) const
+  {
+    ThrowIfError(TRITONSERVER_MetricIncrement(triton_object_, val));
+  }
+
+  void SetValue(double val) const
+  {
+    ThrowIfError(TRITONSERVER_MetricSet(triton_object_, val));
+  }
+
+  TRITONSERVER_MetricKind Kind() const
+  {
+    TRITONSERVER_MetricKind val = TRITONSERVER_METRIC_KIND_COUNTER;
+    ThrowIfError(TRITONSERVER_GetMetricKind(triton_object_, &val));
+    return val;
+  }
+};
+
+// Deferred definitions..
+PyInferenceRequest::PyInferenceRequest(
+    PyServer& server, const std::string& model_name,
+    const int64_t model_version)
+{
+  ThrowIfError(TRITONSERVER_InferenceRequestNew(
+      &triton_object_, server.Ptr(), model_name.c_str(), model_version));
+  owned_ = true;
+}
+
+// [FIXME] module name?
+PYBIND11_MODULE(triton_bindings, m)
```

Comment:
Are there any potentially long running c++ methods where we want to release the GIL or would all calls be  non blocking in any case?